src/broadcast/broadcast.rs

```diff
@@ -200,8 +217,9 @@ impl<N: NodeIdT> Broadcast<N> {
             let fault_kind = FaultKind::ReceivedValueFromNonProposer;
             return Ok(Fault::new(sender_id.clone(), fault_kind).into());
         }
-        if self.echo_sent {
-            if self.echos.get(self.our_id()) == Some(&p) {
+
+        if let Some(EchoContent::Full(proof)) = self.echos.get(self.our_id()) {
+            if *proof == p {
```

Comment:
If we only have an `EchoContent::Hash`, then we can't return from the `if let` statement right? We need to update the `echos` map with the actual full proof. So it would have to be

```
match self.echos.get(self.our_id()) {
    Some(val) if val.hash() != p.root_hash() => {
        return Ok(Fault::new(sender_id.clone(), FaultKind::MultipleValues).into())
    }
    Some(EchoContent::Full(proof)) if *proof == p => {
        warn!(
            "Node {:?} received Value({:?}) multiple times from {:?}.",
            self.our_id(),
            HexProof(&p),
            sender_id
        );
        return Ok(Step::default());
    },
    _ => ()
};
```